File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/Identity.java
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.nd4j.linalg.api.ops.impl.transforms;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * Identity function
 *
 * @author Adam Gibson
 */
public class Identity extends BaseDynamicTransformOp {

    public Identity(SameDiff sd, SDVariable input){
        super(sd, new SDVariable[]{input}, false);
    }

    public Identity(){ }

    @Override
    public String opName() {
        return "identity";
    }

    @Override
    public String onnxName() {
        return "Constant";
    }

    @Override
    public String tensorflowName() {
        return "Identity";
    }

    @Override
    public String[] tensorflowNames() {
        return new String[]{"Identity", "NoOp"};
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> i_v) {
        //TODO can we skip the identity here?
        return Collections.singletonList(sameDiff.identity(i_v.get(0)));
    }

}

File: libnd4j/include/ops/declarable/generic/nn/activations/identity.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
//  @author raver119@gmail.com
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_identity)

#include <ops/declarable/CustomOperations.h>

namespace sd {
namespace ops {
OP_IMPL(identity, 1, 1, true) {
  auto z = OUTPUT_VARIABLE(0);

  if (!block.isInplace()) {
    auto first = INPUT_VARIABLE(0);

    // we hope for memcpy here
    z->assign(first);
  }

  return Status::OK;
}
DECLARE_SYN(linear, identity);

DECLARE_TYPES(identity) { getOpDescriptor()->setAllowedInputTypes(0, ANY)->setSameMode(true); }

OP_IMPL(identity_bp, 2, 1, true) {
  auto first = INPUT_VARIABLE(0);
  auto epsilon = INPUT_VARIABLE(1);
  auto z = OUTPUT_VARIABLE(0);

  z->assign(epsilon);

  return Status::OK;
}
DECLARE_SYN(LinearGrad, identity_bp);

DECLARE_TYPES(identity_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, ANY)
      ->setAllowedInputTypes(1, {ALL_FLOATS})
      ->setAllowedOutputTypes(0, {ALL_FLOATS});
}
}  // namespace ops
}  // namespace sd

#endif
